File: py/three.py
class Node:
    def __init__(self, elem):
        self.elem = elem
        self.count = 1
        self.right = None
        self.left = None


class Three:
    def __init__(self):
        self.root = None

    def add(self, elem):
        if self.root is None:
            self.root = Node(elem)
        else:
            self._insert(self.root, elem)

    def _insert(self, nodo: Node, elem):
        if elem > nodo.elem:
            if nodo.right is None:
                nodo.right = Node(elem)
            else:
                self._insert(nodo.right, elem)
        elif elem < nodo.elem:
            if nodo.left is None:
                nodo.left = Node(elem)
            else:
                self._insert(nodo.left, elem)
        else:
            nodo.count += 1

    def print_right(self):
        return self._print(self.root)

    def _print(self, nodo):
        if nodo is None:
            return
        print(f"elem: {nodo.elem}, count: {nodo.count}")
        if nodo.right:
            self._print(nodo.right)

        if nodo.left:
            self._print(nodo.left)


three = Three()

three.add(1)
three.add(2)
three.add(2)
three.add(0)
three.add(-1)

three.print_right()

File: src/three.rs
use std::cmp::Ordering;

type Link<T> = Box<Node<T>>;

struct Node<T: Ord>
where
    T: Ord + std::fmt::Display,
{
    elem: T,
    right: Option<Link<T>>,
    left: Option<Link<T>>,
    count: u32,
}

pub struct Three<T>
where
    T: Ord + std::fmt::Display,
{
    root: Option<Link<T>>,
}

impl<T> Three<T>
where
    T: Ord + std::fmt::Display,
{
    pub fn new() -> Self {
        Self { root: None }
    }

    pub fn add(&mut self, elem: T) {
        match self.root.as_mut() {
            Some(root) => add_recursive(root, elem),
            None => self.root = Node::new(elem),
        };
    }

    pub fn print(&self) {
        self.print_recursive(self.root.as_ref())
    }

    fn print_recursive(&self, nodo: Option<&Box<Node<T>>>) {
        match nodo {
            Some(nodo) => {
                println!("elem: {}, count: {}", nodo.elem, nodo.count);
                if let Some(right) = nodo.right.as_ref() {
                    self.print_recursive(Some(right));
                }

                if let Some(left) = nodo.left.as_ref() {
                    self.print_recursive(Some(left));
                }
            }
            None => println!("End"),
        }
    }
}

fn add_recursive<T>(nodo: &mut Box<Node<T>>, elem: T)
where
    T: Ord + std::fmt::Display,
{
    match elem.cmp(&nodo.elem) {
        Ordering::Greater => match nodo.right.as_mut() {
            Some(right_nodo) => add_recursive(right_nodo, elem),
            None => nodo.right = Node::new(elem),
        },
        Ordering::Less => match nodo.left.as_mut() {
            Some(left_nodo) => add_recursive(left_nodo, elem),
            None => nodo.left = Node::new(elem),
        },
        Ordering::Equal => {
            nodo.count += 1;
        }
    }
}

impl<T> Node<T>
where
    T: Ord + std::fmt::Display,
{
    fn new(elem: T) -> Option<Box<Self>> {
        Some(Box::new(Self {
            elem,
            left: None,
            right: None,
            count: 1,
        }))
    }
}
